auth/internal/transport/cba.go

```diff
@@ -40,21 +42,90 @@ const (
 	googleAPIUseCertSource = "GOOGLE_API_USE_CLIENT_CERTIFICATE"
 	googleAPIUseMTLS       = "GOOGLE_API_USE_MTLS_ENDPOINT"
 	googleAPIUseMTLSOld    = "GOOGLE_API_USE_MTLS"
+
+	universeDomainPlaceholder = "UNIVERSE_DOMAIN"
 )
 
 var (
-	mdsMTLSAutoConfigSource mtlsConfigSource
+	mdsMTLSAutoConfigSource     mtlsConfigSource
+	errUniverseNotSupportedMTLS = errors.New("mTLS is not supported in any universe other than googleapis.com")
 )
 
 // Options is a struct that is duplicated information from the individual
 // transport packages in order to avoid cyclic deps. It correlates 1:1 with
 // fields on httptransport.Options and grpctransport.Options.
 type Options struct {
-	Endpoint            string
-	DefaultEndpoint     string
-	DefaultMTLSEndpoint string
-	ClientCertProvider  cert.Provider
-	Client              *http.Client
+	Endpoint                string
+	DefaultMTLSEndpoint     string
+	DefaultEndpointTemplate string
+	ClientCertProvider      cert.Provider
+	Client                  *http.Client
+	UniverseDomain          string
+	DefaultUniverseDomain   string
+}
+
+// getDefaultUniverseDomain returns the default service domain for a given Cloud
+// universe, as configured with internaloption.WithDefaultUniverseDomain.
+// The default value is "googleapis.com".
+func (o *Options) getDefaultUniverseDomain() string {
+	if o.DefaultUniverseDomain == "" {
+		return internal.DefaultUniverseDomain
+	}
+	return o.DefaultUniverseDomain
+}
+
+// getUniverseDomain returns the default service domain for a given Cloud
+// universe, as configured with option.WithUniverseDomain.
+// The default value is the value of getDefaultUniverseDomain, as configured
+// with internaloption.WithDefaultUniverseDomain.
+func (o *Options) getUniverseDomain() string {
+	if o.UniverseDomain == "" {
+		return o.getDefaultUniverseDomain()
+	}
+	return o.UniverseDomain
+}
+
+// isUniverseDomainGDU returns true if the universe domain is the default Google
+// universe.
+func (o *Options) isUniverseDomainGDU() bool {
+	return o.getUniverseDomain() == o.getDefaultUniverseDomain()
+}
+
+// defaultEndpoint returns the DefaultEndpointTemplate merged with the
+// Universe Domain if the DefaultEndpointTemplate is set, otherwise returns an
+// empty string.
+func (o *Options) defaultEndpoint() string {
+	if o.DefaultEndpointTemplate == "" {
+		return ""
+	}
+	return strings.Replace(o.DefaultEndpointTemplate, universeDomainPlaceholder, o.getUniverseDomain(), 1)
+}
+
+// mergedEndpoint merges a user-provided Endpoint of format host[:port] with the
+// default endpoint.
+func (o *Options) mergedEndpoint() (string, error) {
+	defaultEndpoint := o.defaultEndpoint()
+	if defaultEndpoint == "" {
```

Comment:
This is not a field that users set. It should be valid to not have any InternalOption sent.

Reply:
Removed, thanks.